### ISSUE TITLE
RNN needs one output for every timestep

### ISSUE THREAD

**Issue by scrcro** [mlpack/mlpack#1286]
I was trying to use the RNN for training a n-gram word prediction model. For my purpose it does not make sense to have an output for every timestep. Perhaps we can have a setting in RNN to dictate the output expectation. I believe Keras provides a parameter called "return_sequence" for the same purpose. Thoughts on how to achieve this?

**Comment by scrcro** [comment]
@zoq  Thank you! I will take a look here and post back.

**Comment by scrcro** [comment]
@zoq I think using `responses.slice(seqNum)` is the problem. Each slice is supposed to be a timestep input. The sequence number is essentially a step in the BPTT. Thus we need to change the `responses.slice(seqNum)` to something else. I have not gone through the complete code. But this is what appears to me to be happening. Please suggest.

**Comment by zoq** [comment]
You are right, my initial idea was to use responses.slice(0) for each time step, if ``sequence`` is set. Another idea is to add another overload for the ``Train`` method, that takes ``arma::cube`` as predictors and ``arma::mat`` as reponses. In the function itself, we can use the responses. The only problem I see is that we would have to hold different responses one for each type.

**Comment by scrcro** [comment]
Thank you. I am slightly against having another method. It would increase confusion, I believe. 
As of now if we just use the first response slice always if `sequence` is set then will it work? 
What is a good set of data to test it? Reber grammar?

**Comment by zoq** [comment]
Reber grammar isn't suitable here, but we could create an artificial dataset, what you think?

**Comment by scrcro** [comment]
I thought so too. Reber grammar may not be good. Let me see if there is a
free dataset that can be used.

**Comment by scrcro** [comment]
How about "Experiment 2" from the [Hochreiter/Schmidthuber paper](http://www.bioinf.jku.at/publications/older/2604.pdf) ?
This seems simple enough.

**Comment by zoq** [comment]
right, I mean as I said we could also just create some artificial data, e.g. a sine generator and assign each a unique label.

**Comment by scrcro** [comment]
Thank you. I have a doubt on this [code](https://github.com/mlpack/mlpack/blob/486be65dab54e8f0fba4066827c8358ffdea3d4e/src/mlpack/methods/ann/rnn_impl.hpp#L314)
What is this one used for? Will replacing `responses.slice(rho - seqNum - 1)` with `responses.slice(0)` work?

**Comment by Manthan-R-Sheth** [comment]
Yes, I think `responses.slice(0)` should work for the case of multi-input and single output RNN model. However, we need not perform `rho` number of steps for the BPTT.  I think just executing the backprop once by considering `rho` input timestamps should be sufficient for updating the weights of the parameters, in such a setting.

**Comment by zoq** [comment]
Agreed, realistically, you could write the test, and see if the correct input/output pair is passed to the network at each step.

**Comment by scrcro** [comment]
Thank you both @zoq @Manthan-R-Sheth . Let me run some tests to see how this works.

**Comment by scrcro** [comment]
Sorry about the rather late response. Was sucked into other things..
@Manthan-R-Sheth  I am not sure what you mean here. I think we still need to run the Backward rho number of times, although the output is only one for rho entries. 
If the network is unrolled shouldn't the Backward be called rho times? Can you please elaborate what you meant? Shouldn't the weights be updated rho times?
I mean line [321 ](https://github.com/mlpack/mlpack/blob/efd29addc8a16704d8aac85ed2994bdc2b4d65e7/src/mlpack/methods/ann/rnn_impl.hpp#L321)and [322 ](https://github.com/mlpack/mlpack/blob/efd29addc8a16704d8aac85ed2994bdc2b4d65e7/src/mlpack/methods/ann/rnn_impl.hpp#L322)of must still be invoked rho times. 

@zoq 
I guess a sine wave test will work too.

**Comment by Manthan-R-Sheth** [comment]
@scrcro Sorry for late response,
For single output , multi input sequences (i.e. sentiment of a sentence), just one BPTT step at the end of the sentence makes sense. 
You can refer [here](https://stats.stackexchange.com/questions/219914/rnns-when-to-apply-bptt-and-or-update-weights?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa).

Also, intuitively, it makes more sense to get the context of the entire sentence first before assigning a sentiment to it (i.e. before performing BPTT step). Let me know what you think.

**Comment by scrcro** [comment]
@Manthan-R-Sheth Got it. But I was talking about a slightly different thing. I am trying to see how the current implementation need to change. I understand that BPTT have to run once, but it needs to include the unfolded layers.

Also another question, is there a way to see/log the optimization progress? Mainly to understand how the net is learning/ or learning if at all?

**Comment by Manthan-R-Sheth** [comment]
You can define the epochs and the number of iterations (depending on batch size) in each epoch. After each batch update, you can keep accumulating the error (by comparing the predictions with true values) and at the end of each epoch you can display the mean of that epoch.
Thus there will be the error displayed after each epoch and you can carry forward your work.

Refer to ForwardBackwardTest of feedforward_network_test.cpp for more clarity.

@scrcro

**Comment by scrcro** [comment]
Thank you, for the pointer. I did find another way to do it. Turns out it is already being logged in the SGD optimizer (at least) all that is required is to use the `ParseCommandLine` function and pass a -v parameter to the main program.

**Comment by Manthan-R-Sheth** [comment]
@scrcro 
Are you planning to make a PR for the same? Else, I can make the basic PR and we can collaborate there?

**Comment by scrcro** [comment]
Yeah. I am testing it right now with a noisy sine wave. If that works I
will submit a PR. I am happy to share what I have done so far.
